Mathlib/CategoryTheory/Limits/Shapes/Pullback/Square.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Square
+import Mathlib.CategoryTheory.Limits.Shapes.Pullback.CommSq
+
+/-!
+# Commutative squares that are pushout or pullback squares
+
+In this file, we translate the `IsPushout` and `IsPullback`
+API for the objects of the category `Square C` of commutative
+squares in a category `C`.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+open Limits
+
+variable {C : Type u} [Category.{v} C]
+
+namespace Square
+
+variable (sq : Square C)
+
+/-- The pullback cone attached to a commutative square. -/
+abbrev pullbackCone : PullbackCone sq.f₂₄ sq.f₃₄ :=
+  PullbackCone.mk sq.f₁₂ sq.f₁₃ sq.fac
+
+/-- The pushout cocone attached to a commutative square. -/
+abbrev pushoutCocone : PushoutCocone sq.f₁₂ sq.f₁₃ :=
+  PushoutCocone.mk sq.f₂₄ sq.f₃₄ sq.fac
+
+/-- The condition that a commutative square is a pullback square. -/
+protected def IsPullback : Prop :=
+  IsPullback sq.f₁₂ sq.f₁₃ sq.f₂₄ sq.f₃₄
+
+/-- The condition that a commutative square is a pushout square. -/
+protected def IsPushout : Prop :=
+  IsPushout sq.f₁₂ sq.f₁₃ sq.f₂₄ sq.f₃₄
+
+lemma isPullback_iff :
+    sq.IsPullback ↔ Nonempty (IsLimit sq.pullbackCone) :=
+  ⟨fun h ↦ ⟨h.isLimit⟩, fun h ↦ { w := sq.fac, isLimit' := h }⟩
+
+lemma isPushout_iff :
+    sq.IsPushout ↔ Nonempty (IsColimit sq.pushoutCocone) :=
+  ⟨fun h ↦ ⟨h.isColimit⟩, fun h ↦ { w := sq.fac, isColimit' := h }⟩
+
+lemma IsPullback.mk (h : IsLimit sq.pullbackCone) : sq.IsPullback :=
+  sq.isPullback_iff.2 ⟨h⟩
+
+lemma IsPushout.mk (h : IsColimit sq.pushoutCocone) : sq.IsPushout :=
+  sq.isPushout_iff.2 ⟨h⟩
+
+variable {sq}
+
+/-- If a commutative square `sq` is a pullback square,
+then `sq.pullbackCone` is limit. -/
+noncomputable nonrec def IsPullback.isLimit (h : sq.IsPullback) :
```

Comment:
Are these `nonrec` necessary? I generally prefer to avoid these in favour of just disambiguating identifiers using more of the namespace. But I wouldn't insist.

Reply:
I have just disambiguated the namespaces...